build.sc

```diff
@@ -41,98 +37,42 @@ trait SourcecodeModule extends PublishModule with MimaCheck {
-      .flatMap(source =>
-        Seq(
-          PathRef(source.path / os.up / "test" / source.path.last),
-          PathRef(source.path / os.up / os.up / "test" / source.path.last)
-        )
-      )
-      .distinct
-  )
-}
+    object test extends ScalaTests{
 
-object sourcecode extends Module {
-  object jvm extends Cross[JvmSourcecodeModule](scalaVersions: _*)
-  class JvmSourcecodeModule(val crossScalaVersion: String)
-    extends SourcecodeMainModule with ScalaModule with SourcecodeModule {
-
-    object test extends SourcecodeTestModule{
-      def scalaVersion = crossScalaVersion
-      def moduleDeps = Seq(JvmSourcecodeModule.this)
-      val crossScalaVersion = JvmSourcecodeModule.this.crossScalaVersion
+      def testFramework = ""
     }
   }
 
-  object js extends Cross[JsSourcecodeModule](scalaJSVersions: _*)
-  class JsSourcecodeModule(val crossScalaVersion: String, crossJSVersion: String)
-    extends SourcecodeMainModule with ScalaJSModule with SourcecodeModule {
-    def offset = os.up
-
-    def scalaJSVersion = crossJSVersion
-    object test extends SourcecodeTestModule with ScalaJSModule{
-      def scalaVersion = crossScalaVersion
-      def scalaJSVersion = crossJSVersion
-      def offset = os.up
-      def moduleDeps = Seq(JsSourcecodeModule.this)
-      val crossScalaVersion = JsSourcecodeModule.this.crossScalaVersion
+  object js extends Cross[JsSourcecodeModule](scalaVersions)
+  trait JsSourcecodeModule extends SourcecodeMainModule with ScalaJSModule with SourcecodeModule {
+
+    def scalaJSVersion = "1.12.0"
+    object test extends ScalaJSTests{
+      def testFramework = ""
     }
   }
 
-  object native extends Cross[NativeSourcecodeModule](scalaNativeVersions: _*)
-  class NativeSourcecodeModule(val crossScalaVersion: String, crossScalaNativeVersion: String)
-    extends SourcecodeMainModule with ScalaNativeModule with SourcecodeModule {
-    def offset = os.up
-
-    def scalaNativeVersion = crossScalaNativeVersion
-
-    override def docJar =
-      if (crossScalaVersion.startsWith("2.")) super.docJar
-      else T {
-        val outDir = T.ctx().dest
-        val javadocDir = outDir / "javadoc"
-        os.makeDir.all(javadocDir)
-        mill.api.Result.Success(mill.modules.Jvm.createJar(Agg(javadocDir))(outDir))
-      }
-
-    object test extends SourcecodeTestModule with ScalaNativeModule{
-      def scalaVersion = crossScalaVersion
-      def scalaNativeVersion = crossScalaNativeVersion
-      def offset = os.up
-      def moduleDeps = Seq(NativeSourcecodeModule.this)
-      val crossScalaVersion = NativeSourcecodeModule.this.crossScalaVersion
+  object native extends Cross[NativeSourcecodeModule](scalaVersions)
+  trait NativeSourcecodeModule extends SourcecodeMainModule with ScalaNativeModule with SourcecodeModule {
+
+    def scalaNativeVersion = "0.5.0"
+    object test extends ScalaNativeTests{
+      // stub to make use of test plumbing but not running a test suite
```

Comment:
Why can't we run tests on Scala Native?

Reply:
@lolgab sourcecode doesnt use a test framework, it just runs through the asserts starting from `def main`. But the build level plumbing for nested test modules w.r.t. cross version and cross platform modules is very handy, so I'm abusing it a bit